feat(behavior) Add Tap-Dance (2022) (#1139)
zmkfirmware/zmk: docs/docs/behaviors/tap-dance.md

---
title: Tap-Dance Behavior
sidebar_label: Tap-Dance
---

## Summary

> A tap-dance key outputs a keycode or behavior corresponding to how many times it is pressed.

Reviewer: Perhaps a minor doc tweak:

```suggestion
A tap-dance key invokes a different behavior (e.g. `kp`) corresponding to how many times it is pressed.
```

Tap-dances are completely custom, so for every unique tap-dance key,
a new tap-dance must be defined in your keymap's `behaviors`.

Tap-dances are designed to resolve immediately when interrupted by another keypress.
Meaning, when a keybind is pressed other than any active tap-dances,
the tap-dance will activate according to the current value of its
counter before the interrupting keybind is registered.

### Configuration

#### `tapping-term-ms`

Defines the maximum elapsed time after the last tap-dance keybind press
before a binding is selected from [`bindings`](#bindings).
Default value is `200`ms.

#### `bindings`

An array of one or more keybinds. This list can include [any ZMK keycode](../codes/) and keybinds for ZMK behaviors.

#### Example Usage

This example configures a tap-dance that outputs the number of keypresses from 1-3.

```
#include <behaviors.dtsi>
#include <dt-bindings/zmk/keys.h>

/ {
	behaviors {

> 		td: tap_dance {

Reviewer: Perhaps number this to reinforce the need for uniqueness:

```suggestion
		td0: tap_dance_0 {
```

            compatible = "zmk,behavior-tap-dance";

>             label = "TAP_TEST";

Reviewer: Ditto:

```suggestion
            label = "TAP_DANCE_0";
```

            #binding-cells = <0>;
            tapping_term_ms = <200>;
            bindings = <&kp N1>, <&kp N2>, <&kp N3>;
        };
	};

	keymap {
		compatible = "zmk,keymap";

		default_layer {
			bindings = <

> 	            &td

Reviewer: ```suggestion
	            &td0
```

			>;
		};
	};
};
```

The following image describes the behavior of this particular tap-dance.

> ![Timing Diagram](../assets/tap-dance/timing_diagram.svg)

Reviewer 2: ![image](https://user-images.githubusercontent.com/7876996/154879709-07c1c126-5d21-496c-a1af-c00842c0d117.png)
I think the intention here is to show an example where the tap dance key is held (blue `td` block) until `j` key is tapped. However this doesn't seem to match the diagram where the blue block cuts off before the orange block starts.

Author: If we read this part of the diagram with only the code in mind, the `td` block doesn't necessarily need to be held down when the `j` is tapped since the tap-dance has technically not been decided yet since the `tapping-term-ms` has not fully elapsed.
However, I definitely see your point about the confusion there since the documentation currently implies that interrupting key-bindings only trigger the tap-dance when the tap-dance key is also held down. Not sure at the moment if adding a note on the image or just putting it directly in the docs would be better. Thanks for the feedback!

Reviewer 2: I see, it makes sense that we call it an "interrupt" also when we are waiting for a potential next tap, and not only while the key is held down (which would be the case for e.g. a hold-tap). I am also not sure how to explain it better; I can add a suggestion if I come up with one.

:::note
Since all `bindings` used for `&td` are [key presses](../behaviors/key-press),
they will release as soon as an interrupting key press occurs. If a modifier key like
`LSHIFT` were to replace the `N1` binding, it would remain pressed until

> `&td`'s binding is released. That is to say, the output would instead be `J`.

Reviewer 2: I am a bit confused about the interrupt behavior as described here -- I wonder if this could be made clearer. 

Modifiers like `LSHIFT` are technically key press behaviors as well, so there is a contradiction with the first sentence. In addition the last two sentences seem to imply there is special logic for modifiers, if that is the case it can be made clearer.

Reviewer 2: After thinking about this one more: I wonder if the above diagram is slightly incorrect with regards to the timing of `j` output? This paragraph makes sense to me if `1` and `j` were output at the same time in the above diagram, rather than `j` being output after `1` is.

Author: I'll check in the serial monitor if they output at the same time. At least visually, the output in say, a text editor, `1` comes before `j`.

Author: ```
[00:00:09.845,397] [0m<dbg> zmk.on_keymap_binding_pressed: position 1 keycode 0x70004[0m
[00:00:09.845,428] [0m<dbg> zmk.hid_listener_keycode_pressed: usage_page 0x07 keycode 0x04 implicit_mods 0x00 explicit_mods 0x00[0m
[00:00:09.845,428] [0m<dbg> zmk.zmk_hid_implicit_modifiers_press: Modifiers set to 0x00[0m
[00:00:09.845,428] [0m<dbg> zmk.zmk_endpoints_send_report: usage page 0x07[0m
[00:00:09.845,489] [0m<dbg> zmk.release_captured_events: Releasing key position event for position 0 pressed[0m
[00:00:09.845,520] [0m<dbg> zmk.position_state_changed_listener: 0 bubble (no undecided hold_tap active)[0m
[00:00:09.845,520] [0m<dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 0, binding name: KEY_PRESS[0m
[00:00:09.845,550] [0m<dbg> zmk.on_keymap_binding_pressed: position 0 keycode 0x70027[0m

```
They output on the same millisecond, but `1` definitely comes before `j`, as expected. Will update the timing diagram to reflect this.

Author: ![Draft 4 (Main)@1 5x](https://user-images.githubusercontent.com/27038994/155202991-279d4db0-7aa8-46bf-8f41-89b9ff84db6c.png)

Made some small tweaks to the timing diagram according to your suggestions. If you have any further comments, feel free to share them.

Author: I realize that the `tap, tap, tap` is missing an additional extension to the dashed red line indicating the cancelled tapping-term from the third press of `td`. This updated image would be more accurate.
![Draft 4 (Main)@1 5x](https://user-images.githubusercontent.com/27038994/155205434-702a96ea-d736-4654-895e-bf2718016ac4.png)

Reviewer 2: > Made some small tweaks to the timing diagram according to your suggestions. If you have any further comments, feel free to share them.

This looks good to me, thanks!

> I realize that the tap, tap, tap is missing an additional extension to the dashed red line indicating the cancelled tapping-term from the third press of td. This updated image would be more accurate.

I see, tapping term gets refreshed after the second tap's press so this shows that extension. I assume all timers are killed after the third tap anyway since tap-dance only has three items: I guess it is still good to illustrate the extension in this case.

Author: > I am a bit confused about the interrupt behavior as described here -- I wonder if this could be made clearer.
> 
> Modifiers like `LSHIFT` are technically key press behaviors as well, so there is a contradiction with the first sentence. In addition the last two sentences seem to imply there is special logic for modifiers, if that is the case we should state it explicitly.

```
Alphanumeric [`key press`](key-press.md) bindings, like those used for `td0`,
will release as soon as an interrupting alphanumeric key press occurs.
If a modifier key like `LSHIFT` were to replace the `N1` binding, it would
remain pressed until `td0`'s binding is released.
That is to say, the output would instead be `J` and any following alphanumeric
key presses would be capitalized as long as `&td0` is held down.
```
Just got around to addressing your concerns regarding the wording surrounding key presses too. How does this look?

Reviewer 2: I think that's better, thanks! In "as soon as an interrupting alphanumeric key press occurs", is the fact that the interrupting key press is alphanumeric important? If not we can remove that to simplify. Also it would be good to point out that this example refers to the last case in the diagram, e.g. we can augment the last part to be

> For instance, if a modifier key like `LSHIFT` were to replace the `N1` binding in the last example above, it would
remain pressed until `td0`'s binding is released and the output would instead be `J`. Any following alphanumeric
key presses would also be shifted while `&td0` is held down.

Author: Specification of the alphanumeric nature of the keypress is important because like you said previously, modifiers are technically keypresses too. I also realized that due to the limited scope of our findings this far, I'm not sure if other types of behavior bindings interrupt tap-dances, so that might be something for other testers to discover as well.

Author: Actually, I think I see your point a little better. I used "alphanumeric" twice in the same sentence which reads a little weird. We probably don't need the second copy of the word for readability.

---
title: Tap-Dance Behavior
sidebar_label: Tap-Dance
---

## Summary

> A tap-dance key invokes a different behavior (e.g. `kp`) corresponding to how many times it is pressed.
> Tap-dances are completely custom, so for every unique tap-dance key,
> a new tap-dance must be defined in your keymap's `behaviors`.

Reviewer 3: Nice work! 

I think a brief example could be useful here, before getting into the technical stuff:

```suggestion
A tap-dance key invokes a different behavior (e.g. `kp`) corresponding to how many times it is pressed. For example, you could configure a tap-dance key that toggles [Caps Word](../caps-word) if tapped once, or Caps _Lock_ if tapped twice.

Tap-dances are completely custom, so for every unique tap-dance key,
a new tap-dance must be defined in your keymap's `behaviors`.
```

I'm not sure if that's the best example, but I tried to think of something practical for those who might not understand the potential use cases of tap-dance.

Author: > For example, you could configure a tap-dance key that toggles
> [Caps Word](../caps-word) if tapped once, or Caps _Lock_ if tapped twice.
> The expandability of the number of [`bindings`](#bindings) attached to a
> particular tap-dance is a great way to add more functionality to a single key,
> especially for keyboards with a limited number of keys.

Expanding upon your work here with an extra qualitative statement before the technical stuff
